cairo/src/runner.rs

```diff
@@ -0,0 +1,472 @@
+pub struct CairoVariables<F: FftField> {
+    /// Destination
+    dst: Option<F>,
+    /// First operand
+    op0: Option<F>,
+    /// Second operand
+    op1: Option<F>,
+    /// Result
+    res: Option<F>,
+    /// Destination address
+    dst_addr: F,
+    /// First operand address
+    op0_addr: F,
+    /// Second operand address
+    op1_addr: F,
+    /// Size of the instruction
+    size: F,
+}
+
+impl<F: FftField> CairoVariables<F> {
+    /// This function creates an instance of a default CairoVariables struct
+    pub fn new() -> CairoVariables<F> {
+        CairoVariables {
+            dst: None,
+            op0: None,
+            op1: None,
+            res: None,
+            dst_addr: F::zero(),
+            op0_addr: F::zero(),
+            op1_addr: F::zero(),
+            size: F::zero(),
+        }
+    }
+}
+impl<F: FftField> Default for CairoVariables<F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A data structure to store a current step of Cairo computation
+pub struct CairoStep<'a, F: FftField> {
+    /// current word of the program
+    pub mem: &'a mut CairoMemory<F>,
+    // comment instr for efficiency
+    /// current pointers
+    curr: CairoPointers<F>,
+    /// (if any) next pointers
+    next: Option<CairoPointers<F>>,
+    /// state auxiliary variables
+    vars: CairoVariables<F>,
+}
+
+impl<'a, F: FftField> CairoStep<'a, F> {
+    /// Creates a new Cairo execution step from a step index, a Cairo word, and current pointers
+    pub fn new(mem: &mut CairoMemory<F>, ptrs: CairoPointers<F>) -> CairoStep<F> {
+        CairoStep {
+            mem,
+            curr: ptrs,
+            next: None,
+            vars: CairoVariables::new(),
+        }
+    }
+
+    /// Executes a Cairo step from the current registers
+    pub fn execute(&mut self) {
+        // This order is important in order to allocate the memory in time
+        self.set_op0();
+        self.set_op1();
+        self.set_res();
+        self.set_dst();
+        // If the Option<> thing is not a guarantee for continuation of the program, we may be removing this
+        let next_pc = self.next_pc();
+        let (next_ap, next_fp) = self.next_apfp();
+        self.next = Some(CairoPointers::new(
+            next_pc.unwrap(),
+            next_ap.unwrap(),
+            next_fp.unwrap(),
+        ));
+    }
+
+    /// This function returns the current word instruction being executed
+    pub fn instr(&mut self) -> CairoWord<F> {
+        CairoWord::new(self.mem.read(self.curr.pc).unwrap())
+    }
+
+    /// This function computes the first operand address
+    pub fn set_op0(&mut self) {
+        let reg = {
+            match self.instr().op0_reg() {
+                OP0_AP => self.curr.ap, // reads first word from allocated memory
+                _ => self.curr.fp,      // reads first word from input stack
+            } // no more values than 0 and 1 because op0_reg is one bit
+        };
```

Comment:
```rust
let reg = match self.instr().op0_reg() {
...
};
```

Reply:
And the same for other functions of the module, true